### PR TITLE
Add cloud storage support for Models & Worlds

### PR DESCRIPTION
**Pull request by marcoshuck**
This PR adds support for downloading resources (models and worlds) from S3. Fuel server will drastically increase its throughput by adding support for downloading from S3, this will decrease the load in the main server, and delegate the responsibility to AWS.

It also includes a tool for migrating resources from the host machine mounting an EFS drive to a storage like S3. 

Support for other storage services can be added in future iterations of https://github.com/gazebo-web/gz-go/pull/11.

### CODE REVIEW

File: bundles/common_resources/file_resources.go [hunk 1]
```diff
@@ -210,7 +210,7 @@ func Remove(tx *gorm.DB, res Resource, user string) *gz.ErrMsg {
 // ZipResourceTip creates a new zip file for the given resource. Returns the zip
 // Filesize or an error.
```
Comment (AlejoAsd) [inline bundles/common_resources/file_resources.go]: Please update this to reflect returned values.

File: bundles/models/model.go [hunk 2]
```diff
@@ -148,6 +149,10 @@ func (m *Model) GetUUID() *string {
 	return m.UUID
 }
 
+func (m *Model) toStorage(v uint64) storage.Resource {
+	return storage.NewResource(*m.GetUUID(), storage.KindModels, *m.GetOwner(), v)
+}
+
```
Comment (AlejoAsd) [inline bundles/models/model.go]: I'd like to go back to a previous discussion of not using class methods to convert to types from other packages as it binds packages together.

Instead, consider making this a function that receives a `Model` pointer + parameters and returns a `storage.Resource`.

File: bundles/models/models_service.go [hunk 3]
```diff
@@ -465,8 +468,17 @@ func (ms *Service) DownloadZip(ctx context.Context, tx *gorm.DB, owner, modelNam
 	if errorMsg != nil {
 		return nil, nil, 0, errorMsg
 	}
-	path, resolvedVersion, em := res.GetZip(ctx, model, "models", version)
-	return model, path, resolvedVersion, em
+
+	//_, resolvedVersion, em := res.GetZip(ctx, model, "models", version)
```
Comment (AlejoAsd) [inline bundles/models/models_service.go]: Should this be removed?

File: bundles/models/models_service.go [hunk 4]
```diff
@@ -465,8 +468,17 @@ func (ms *Service) DownloadZip(ctx context.Context, tx *gorm.DB, owner, modelNam
...
+	_, resolvedVersion, em := res.GetRevisionFromVersion(ctx, model, version)
```
Comment (AlejoAsd) [inline bundles/models/models_service.go:472]: I'm assuming that we will keep using Git repos under the hood to at least keep versioning things, even if we don't actually use it to store resources.

File: application.go [hunk 5]
```diff
@@ -216,16 +225,42 @@ func init() {
 			if useAwsInTests {
 				awsBucketEnvVar += "_TEST"
 			}
+			p, err := gz.ReadEnvVar("S3_BUCKET")
+			if err != nil {
+				panic("error reading bucket name")
+			}
```
Comment (AlejoAsd) [inline application.go]: Since we're using an `S3mem` server, consider using some default bucket name instead of failing when running tests.

File: application.go [hunk 6]
```diff
@@ -216,16 +225,42 @@ func init() {
...
 			}
+			globals.SessionS3 = session.Must(session.NewSession())
+			globals.S3 = s3.New(globals.SessionS3)
+			globals.UploaderS3 = s3manager.NewUploader(globals.SessionS3)
+			globals.BucketS3 = p
 			subt.BucketServerImpl = subt.NewS3Bucket(p)
 		}
 	}
 
+	globals.CloudStorage = storage.NewS3v1(globals.S3, globals.UploaderS3, globals.BucketS3)
```
Comment (AlejoAsd) [inline application.go]: Why not just call this `Storage`? It just happens that we're now moving to cloud storage but the implementation doesn't really matter.

File: bundles/subt/bucket_server.go [hunk 7]
```diff
@@ -23,9 +23,9 @@ type S3Bucket struct {
 	// each time the Session is created. Sharing the Session value across all of
 	// your service clients will ensure the configuration is loaded the fewest
 	// number of times possible.
-	sess *session.Session
+	Sess *session.Session
 	// s3 clients are safe to use concurrently.
-	svc    *s3.S3
+	Client *s3.S3
```
Comment (AlejoAsd) [inline bundles/subt/bucket_server.go]: Why were these made public? I can't find any usages that require it.

File: bundles/collections/collections_service.go [hunk 8]
```diff
@@ -158,7 +158,7 @@ func populateCollectionThumbnails(tx *gorm.DB,
 		var r res.Resource
 		var em *gz.ErrMsg
 		if a.Type == TModel {
-			s := &models.Service{}
+			s := &models.Service{Storage: globals.CloudStorage}
 			r, em = s.GetModel(tx, a.AssetOwner, a.AssetName, user)
```
Comment (AlejoAsd) [inline bundles/collections/collections_service.go:162]: We discussed having the services be global. Did you find a reason not to do it or are you trying to minimize the number of changes?

File: globals/globals.go [hunk 9]
```diff
@@ -65,3 +70,15 @@ var MaxCategoriesPerModel int
...
+
+var S3 *s3.S3
+
+var UploaderS3 *s3manager.Uploader
+
+var BucketS3 string
+
+var HTTPTestS3Server *httptest.Server
+
+var CloudStorage storage.Storage
```
Comment (AlejoAsd) [inline globals/globals.go]: Please document these fields.

File: handler_models.go [hunk 10]
```diff
@@ -235,20 +235,20 @@ func ModelOwnerVersionZip(owner, name string, user *users.User, tx *gorm.DB,
...
 	if em != nil {
 		return nil, em
 	}
 
-	zipFileName := fmt.Sprintf("model-%sv%d.zip", *model.UUID, ver)
+	//zipFileName := fmt.Sprintf("model-%sv%d.zip", *model.UUID, ver)
 
 	// Remove request header to always serve fresh
-	r.Header.Del("If-Modified-Since")
+	//r.Header.Del("If-Modified-Since")
```
Comment (AlejoAsd) [inline handler_models.go]: Should this be removed?

```suggestion
```

File: handler_models.go [hunk 11]
```diff
@@ -235,20 +235,20 @@ func ModelOwnerVersionZip(owner, name string, user *users.User, tx *gorm.DB,
...
-	model, zipPath, ver, em := (&models.Service{}).DownloadZip(r.Context(), tx,
+	svc := &models.Service{Storage: globals.CloudStorage}
+	_, zipPath, ver, em := svc.DownloadZip(r.Context(), tx,
 		owner, name, modelVersion, user, r.UserAgent())
 	if em != nil {
 		return nil, em
 	}
 
-	zipFileName := fmt.Sprintf("model-%sv%d.zip", *model.UUID, ver)
+	//zipFileName := fmt.Sprintf("model-%sv%d.zip", *model.UUID, ver)
```
Comment (AlejoAsd) [inline handler_models.go]: Should this be removed?

```suggestion
```

File: handler_models.go [hunk 12]
```diff
@@ -235,20 +235,20 @@ func ModelOwnerVersionZip(owner, name string, user *users.User, tx *gorm.DB,
...
 
 	// Remove request header to always serve fresh
-	r.Header.Del("If-Modified-Since")
+	//r.Header.Del("If-Modified-Since")
 	// Set zip response headers
 	w.Header().Set("Content-Type", "application/zip")
-	w.Header().Set("Content-Disposition", fmt.Sprintf("attachment; filename=\"%s\"", zipFileName))
+	//w.Header().Set("Content-Disposition", fmt.Sprintf("attachment; filename=\"%s\"", zipFileName))
```
Comment (AlejoAsd) [inline handler_models.go]: Why is this being removed?

**Review by AlejoAsd — CHANGES_REQUESTED** [review]

File: bundles/collections/collections_service.go (hunk 8 above)
Comment (marcoshuck) [reply bundles/collections/collections_service.go:162]: I'm trying to minimize the number of changes, but thanks for following up on this.

File: bundles/models/model.go (hunk 2 above)
Comment (marcoshuck) [reply bundles/models/model.go]: > I'd like to go back to a previous discussion of not using class methods to convert to types from other packages as it binds packages together.

I could argue that's what we indirectly do with `gorm.Model`, but well, let's use a function then!

Comment (AlejoAsd) [reply bundles/models/model.go]: We do, but that doesn't make it right and we don't really have a choice because of the way `gorm` works.

File: handler_models.go (hunk 12 above)
Comment (marcoshuck) [reply handler_models.go]: I assume that the S3 server handles all the HTTP headers, I kept them as comments so we can discuss it in the PR.

File: bundles/models/model.go [hunk 13]
```diff
@@ -149,8 +149,9 @@ func (m *Model) GetUUID() *string {
 	return m.UUID
 }
 
-func (m *Model) toStorage(v uint64) storage.Resource {
-	return storage.NewResource(*m.GetUUID(), storage.KindModels, *m.GetOwner(), v)
+// newModelStorageResource creates a new storage.Resource for the given model associated to the given version.
+func newModelStorageResource(m *Model, v uint64) storage.Resource {
```
Comment (AlejoAsd) [inline bundles/models/model.go]: I think `new` in the function name is implicit here, and explicitly mentioned in the documentation. Unless there's any naming collisions, we can drop it.

File: globals/globals.go [hunk 14]
```diff
@@ -71,14 +71,20 @@ var MaxCategoriesPerModel int
 // ElasticSearch is a pointer to the Elastic Search client.
 var ElasticSearch *elasticsearch.Client
 
+// SessionS3 contains a AWS session handler.
```
Comment (AlejoAsd) [inline globals/globals.go]:
```suggestion
// SessionS3 contains an AWS session used to perform S3 operations.
```

File: application.go [hunk 15]
```diff
@@ -225,11 +225,7 @@ func init() {
 			if useAwsInTests {
 				awsBucketEnvVar += "_TEST"
```
Comment (AlejoAsd) [inline application.go]: The warning here is happening because this code is only triggered if `isGoTest` is true, but the `awsBucketEnvVar` variable is only used if `isGoTest` is false.

File: handler_models.go [hunk 16]
```diff
@@ -235,20 +235,19 @@ func ModelOwnerVersionZip(owner, name string, user *users.User, tx *gorm.DB,
...
 
-	//zipFileName := fmt.Sprintf("model-%sv%d.zip", *model.UUID, ver)
-
-	// Remove request header to always serve fresh
-	//r.Header.Del("If-Modified-Since")
 	// Set zip response headers
+	zipFileName := fmt.Sprintf("%d.zip", ver)
 	w.Header().Set("Content-Type", "application/zip")
-	//w.Header().Set("Content-Disposition", fmt.Sprintf("attachment; filename=\"%s\"", zipFileName))
+	r.Header.Del("If-Modified-Since")
```
Comment (AlejoAsd) [inline handler_models.go]: Let's revert this change and move the header deletion back up where it was and include the comment explaining why we're doing this.

File: handler_models.go [hunk 17]
```diff
@@ -257,14 +256,13 @@ func ModelOwnerVersionZip(owner, name string, user *users.User, tx *gorm.DB,
 	// commit the DB transaction
 	// Note: we commit the TX here on purpose, to be able to detect DB errors
 	// before writing "data" to ResponseWriter. Once you write data (not headers)
-	// into it the status code is set to 200 (OK).
```
Comment (AlejoAsd) [inline handler_models.go:260]: Why was this removed?

File: handler_models.go [hunk 18]
```diff
@@ -257,14 +256,13 @@ func ModelOwnerVersionZip(owner, name string, user *users.User, tx *gorm.DB,
...
 	if err := tx.Commit().Error; err != nil {
 		return nil, gz.NewErrorMessageWithBase(gz.ErrorZipNotAvailable, err)
 	}
 
 	// Redirect to the cloud storage
-	http.Redirect(w, r, *zipPath, http.StatusOK)
-	return nil, nil
+	http.Redirect(w, r, *zipPath, http.StatusSeeOther)
```
Comment (AlejoAsd) [inline handler_models.go]: Consider using `302 http.StatusFound` instead of `303 http.StatusSeeOther`.

While they are functionally identical for our use case, the 303 code implies that the request is being redirected to another page to confirm/handle the request and access the resource, whereas the 302 request tells the client that the resource is in the returned address.

**Review by AlejoAsd — CHANGES_REQUESTED** [review]

File: application.go (hunk 15 above)
Comment (marcoshuck) [reply application.go]: Yeah, we can drop this env var.

File: bundles/models/model.go (hunk 13 above)
Comment (marcoshuck) [reply bundles/models/model.go]: I usually prefer keeping function names as operations: do, create, update. I'd prefer keeping the `new` to follow that rule, although I'm just aiming for consistency, we can change it if you want.

Comment (AlejoAsd) [reply bundles/models/model.go]: How about `modelToStorageResource`? Or (even more verbose) `castModelToStorageResource`?

File: handler_models.go [hunk 19]
```diff
@@ -235,20 +235,22 @@ func ModelOwnerVersionZip(owner, name string, user *users.User, tx *gorm.DB,
...
-	zipFileName := fmt.Sprintf("model-%sv%d.zip", *model.UUID, ver)
+	// Set zip response headers
+	zipFileName := fmt.Sprintf("%d.zip", ver)
 
 	// Remove request header to always serve fresh
 	r.Header.Del("If-Modified-Since")
-	// Set zip response headers
+
 	w.Header().Set("Content-Type", "application/zip")
 	w.Header().Set("Content-Disposition", fmt.Sprintf("attachment; filename=\"%s\"", zipFileName))
```
Comment (AlejoAsd) [inline handler_models.go]: On second thought, I believe these no longer apply now that we're returning a `3xx` redirect and not providing a file anymore. We should remove them.

File: handler_models.go [hunk 20]
```diff
@@ -257,15 +259,14 @@ func ModelOwnerVersionZip(owner, name string, user *users.User, tx *gorm.DB,
...
+	// into it the status code is set to 302 (Found).
 	if err := tx.Commit().Error; err != nil {
 		return nil, gz.NewErrorMessageWithBase(gz.ErrorZipNotAvailable, err)
 	}
 
-	// Serve the zip file contents
-	// Note: ServeFile should be always last line, after all headers were set.
-	http.ServeFile(w, r, *zipPath)
-	return nil, nil
+	// Redirect to the cloud storage
```
Comment (AlejoAsd) [inline handler_models.go]: Since the storage can change, let's keep it generic.

```suggestion
	// Redirect to the storage containing the file for download.
```

File: bundles/models/model.go (hunk 13 above)
Comment (marcoshuck) [reply bundles/models/model.go]: What about `parse...`?

File: handler_models.go (hunk 19 above)
Comment (marcoshuck) [reply handler_models.go]: let me try downloading a model from S3 directly and see if the headers are included.

Comment (marcoshuck) [reply handler_models.go]: They're not, we can remove these headers.

File: bundles/models/model.go (hunk 13 above)
Comment (AlejoAsd) [reply bundles/models/model.go]: `parse` makes me thing that you are converting from text into a language representation. I know we've used `parse` in the past and I believe I raised the same point. Up to you.

Comment (marcoshuck) [reply bundles/models/model.go]: I'm going with `cast`, and moved the helper func to the `common_resources` folder where it can be used with the `Resource` interface.

**Review by AlejoAsd — APPROVED** [review]

File: handler_models.go [hunk 21]
```diff
@@ -235,37 +235,24 @@ func ModelOwnerVersionZip(owner, name string, user *users.User, tx *gorm.DB,
...
-	}
-
 	// commit the DB transaction
 	// Note: we commit the TX here on purpose, to be able to detect DB errors
 	// before writing "data" to ResponseWriter. Once you write data (not headers)
-	// into it the status code is set to 200 (OK).
+	// into it the status code is set to 302 (Found).
 	if err := tx.Commit().Error; err != nil {
 		return nil, gz.NewErrorMessageWithBase(gz.ErrorZipNotAvailable, err)
 	}
```
Comment (AlejoAsd) [inline handler_models.go:251]:
```suggestion

	r.Header.Del("Authorization")
```

Comment (german-e-mas) [reply handler_models.go:251]: Urgh, thanks Github UI. No idea why it made a commit. Sorry about that @marcoshuck , I'll revert the change I added.

+1 to this - Browsers automatically pass through the headers even if the location is a different host.

To this, we should add `r.Header.Del("Private-Token")` (check the private token header to see it matches correctly)

File: handler_models.go [hunk 22]
```diff
@@ -235,36 +229,32 @@ func ModelOwnerVersionZip(owner, name string, user *users.User, tx *gorm.DB,
...
+	// before writing "data" to ResponseWriter.
 	if err := tx.Commit().Error; err != nil {
 		return nil, gz.NewErrorMessageWithBase(gz.ErrorZipNotAvailable, err)
 	}
 
-	// Serve the zip file contents
-	// Note: ServeFile should be always last line, after all headers were set.
-	http.ServeFile(w, r, *zipPath)
+	// If ?link=true, fuel will return a link to a cloud storage where the client can perform a subsequent request
+	// to download the resource. If ?link=false or if it is not included, it will serve the file directly to the client.
```
Comment (AlejoAsd) [inline handler_models.go]: Adding the changes included in the `handler_worlds.go` file.

```suggestion
	// If a link was requested, fuel will return a link to a cloud storage where the client can perform a subsequent request
	// to download the resource. If a link was not requested or if it is not included, it will serve the file directly to the client.
```

**Review by AlejoAsd — APPROVED** [review]

File: handler_models.go (hunk 22 above)
Comment (marcoshuck) [reply handler_models.go]: Thanks, I forgot to reflect the change, good catch!

**Comment by marcoshuck** [comment]
Waiting for a new release of gz-go to release this new feature.
